tests/frontend/test_program.py

```diff
@@ -0,0 +1,308 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""Unit tests for program.py"""
+import pytest
+
+pytestmark = pytest.mark.frontend
+
+import numpy as np
+
+import strawberryfields as sf
+
+from strawberryfields import program
+from strawberryfields import ops
+
+
+# make test deterministic
+np.random.random(42)
+A = np.random.random()
+
+
+# all single-mode gates with at least one parameter
+single_mode_gates = [x for x in ops.one_args_gates + ops.two_args_gates if x.ns == 1]
+
+
+@pytest.fixture
+def eng(backend):
+    """Engine fixture."""
+    return sf.Engine(backend)
+
+
+@pytest.fixture
+def prog():
+    """Program fixture."""
+    return sf.Program(2)
+
+
+@pytest.fixture
+def permute_gates():
+    """Returns a list containing an instance of each
+    single mode gate, with random parameters"""
+    params = np.random.random(len(single_mode_gates))
+    gates = [G(p) for G, p in zip(single_mode_gates, params)]
+
+    # random permutation
+    return np.random.permutation(gates)
+
+
+class TestProgram:
+    """Tests the Program class."""
+
+    def test_with_block(self, prog):
+        """Gate application using a with block."""
+        # command queue is empty
+        assert len(prog) == 0
+
+        with prog.context as q:
+            ops.Dgate(0.5) | q[0]
+        # now there is one gate in the queue
+        assert len(prog) == 1
+
+        with prog.context as q:
+            ops.BSgate(0.5, 0.3) | (q[1], q[0])
+        assert len(prog) == 2
+
+    def test_identity_command(self, prog):
+        """Tests that the None command acts as the identity"""
+        identity = program.Command(None, prog.register[0])
+        prog.circuit.append(identity)
+        assert len(prog) == 1
+        prog = prog.compile("base")
+        assert len(prog) == 0
+
+    def test_parent_program(self):
+        """Continuing one program with another."""
+        D = ops.Dgate(0.5)
+        prog = sf.Program(3)
+        with prog.context as q:
+            D | q[1]
+            ops.Del | q[0]
+        cont = sf.Program(prog)
+        with cont.context as q:
+            D | q[0]
+            r = ops.New(1)
+            D | r
+        assert cont.can_follow(prog)
+        assert prog.reg_refs == cont.init_reg_refs
+        assert prog.unused_indices == cont.init_unused_indices
+
+    def test_print_commands(self, eng, prog):
+        """Program.print and Engine.print_applied return correct strings."""
+        prog = sf.Program(2)
+        res = []
+        print_fn = lambda x: res.append(x.__str__())
```

Comment:
our usage of `res` and `print_fn` is a bit confusing to follow in the tests; it's not easy to follow the flow. Perhaps some brief explanatory comment could be useful here

Reply:
done